### ISSUE TITLE
Discard for lambda parameters (_, _) =>

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#111]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See
- https://github.com/dotnet/roslyn/issues/16255

**Comment by DavidArno** [comment]
Please, please, please could this apply to methods too? Sometimes, the parameters must be there, but just aren't used. By way of example, [this Unit type](https://github.com/DavidArno/SuccincT/blob/master/SuccincT/Functional/Unit.cs) has to have both R# and `CA1801` suppressions for the operator overloads:
````cs
public static bool operator ==(Unit u1, Unit u2) => true;
public static bool operator !=(Unit u1, Unit u2) => false;
````
Being able to write those with discards to reassure those suppressions that the parameters must be there, but aren't used, would clean up such code nicely:
````cs
public static bool operator ==(Unit _, Unit _) => true;
public static bool operator !=(Unit _, Unit _) => false;
````

**Comment by jnm2** [comment]
@DavidArno I believe the .NET Frameowkr design guidelines have always been to use the names `left` and `right` specifically.

**Comment by DavidArno** [comment]
@jnm2,

Oh, I didn't know that. I'll give that a try to see if it makes the compiler happier. Would still like to be able to use discards though.

**Comment by jnm2** [comment]
@DavidArno It'll only make the compiler happier if you have code quality being analyzed.

**Comment by acple** [comment]
this proposal also permit in nested lambda expression `_ => _ => true` ?
```csharp
Enumerable.Range(0, 3)
    .SelectMany(_ => Enumerable.Range(0, 3)
        .SelectMany(_ => Enumerable.Range(0, 3))); // i don't need parameters in nested
```

**Comment by gafter** [comment]
@acple Possibly not, unless we also do something like https://github.com/dotnet/roslyn/issues/15793 at the same time.

**Comment by acple** [comment]
@gafter Thank you. in my opinion, lambda parameter shadowing is a bit dangerous but, discarded parameters are safe to overcover. so that should support only with discarded parameters. is this shadowing issue? or discarding issue?

**Comment by gafter** [comment]
@acple This is a shadowing issue. When there is only one lambda parameter named `_`, it is a normal identifier that declares a parameter. It must be so for backward compatibility.

**Comment by acple** [comment]
@gafter okay I understand. desire the identifier definition rules will be relaxed partially... thanks for the information! :smile:

**Comment by MovGP0** [comment]
there should also be a compiler warning/error when using an `_` named variable: 
```csharp
items.ForEach(_ => Console.WriteLine(_));
```

**Comment by DavidArno** [comment]
@MovGP0,

Unfortunately, that would be a breaking change. An analyzer could be added to this affect though.

**Comment by jnm2** [comment]
I think use of a single `_`-named parameter should be encouraged! I don't intend to use it any less.

**Comment by DavidArno** [comment]
@jnm2,

I disagree. If I was going to use the parameter in the lambda body, then I'd use eg `x`. I think `_` should only be used when the parameter(s) aren't used in the body, ie as discards.

I appreciate that's personal opinion though.

**Comment by jnm2** [comment]
@DavidArno Exactly, it is personal opinion. My rationale is that I'd like to not have to specify a parameter at all so until I can do `.Where(@.Foo != null)` or similar, `_ => _.Foo` seems to be the nearest I can get to not specifying a parameter name.

**Comment by MovGP0** [comment]
I disagree that it is a personal opinion, since it is a common convention in functional languages that `_` is a 'throwaway' value that is not to be used. It confuses developers who are used to the convention. 

Thankfully it is quite simple to give it a proper name.

**Comment by gafter** [comment]
> I disagree that it is a personal opinion...

Oh, the irony.

**Comment by firelizzard18** [comment]
Nested discards (`_ => _ => <expr>`) should be supported on at least an opt-in basis.

**Comment by OJacot-Descombes** [comment]
Since a underscore is a valid identifier, shouldn't we use another character as discard character? I suggest the minus sign.

`(-,-) =>` or `- =>`

**Comment by wanton7** [comment]
If I remember correctly also nameof was a valid indentifier before it was added. For consistency _ should be used every for discards.  When this is added I would expect a compile error if _ is accessed when C# target language version is same or greater than version where this was added.

**Comment by HaloFour** [comment]
@wanton7 

> If I remember correctly also nameof was a valid indentifier before it was added. 

It still is.  The `nameof` identifier only resolves to the keyword when there is no other identifier by that name in use:

```cs
static class Program {
    static string nameof(object o) => "tricksy!";

    static void Main() {
        string s = "s";
        Console.WriteLine(nameof(s)); // prints "tricksy!", not "s"
    }
}
```

Similarly, the C# compiler treats `_` as a valid identifier when such an identifier is in scope, but as a "discard" when it's not, except with new syntax that could not be broken by always treated the identifier as a discard:

```cs
string s = "123";
{
    int _;
    int.TryParse(s, out _); // _ is a variable
}
{
    int.TryParse(s, out _); // _ is a discard
}
{
    int _;
    int.TryParse(s, out var _); // _ is always a discard
}
```

This is quite unlike Java where they did fully deprecate `_` as a valid identifier in version 9, and they did fully deprecate `var` as a valid class name in version 10.

The C# team is pretty obsessed with backward compatibility, which is usually a great thing, except (IMHO) here where `_` can mean any number of things and an intent to discard might overwrite a variable or field that happens to have an unfortunate name.

**Comment by firelizzard18** [comment]
@OJacot-Descombes I would prefer `~` as a new discard keyword: `~ => { }`, `(~, ~) => {}`

**Comment by firelizzard18** [comment]
@wanton7 I was responding to a previous comment by @OJacot-Descombes:

> Since a underscore is a valid identifier, shouldn't we use another character as discard character? I suggest the minus sign.

I was adding my two cents to that, saying that if the language designers agree with that comment, I think `~` would be better than `-` (as a universal discard keyword, not as something special for lambdas).

**Comment by tpetrina** [comment]
But still, `(,) => ...` is easier. You know what is even easier? Ensuring `() => ...` also works when you don't care about arguments...

**Comment by theunrepentantgeek** [comment]
> Ensuring () => ... also works when you don't care about arguments...

This would make the language ambiguous.

```
void Volley(int i) { ... }
void Fling(Action action) { ... }
void Fling(Action<int> action) { ... }

void Demo 
{
    Fling(()=> Volley(4);
}
```
If `() =>` included the concept of discards, which overload of `Fling()` would be called?

Also, given that this is currently unambiguous (calling `Fling(Action)`), widening the definition of `()=>` to include implicit discards would break existing code.

**Comment by tpetrina** [comment]
In that case, fail with error ambiguous. It's not like we don't have errors in case of ambiguity.

**Comment by HaloFour** [comment]
@tpetrina 

> In that case, fail with error ambiguous. It's not like we don't have errors in case of ambiguity.

But not in that case.  `()` already legally means "no parameters".  Changing that would break existing code which would make it a non-starter.

**Comment by tpetrina** [comment]
Then you can use `(,,,) => {}` syntax. I actually like using`_` as an name sometimes.

**Comment by DavidArno** [comment]
@tpetrina, using `_` an a name, rather than a discard, is just plain weird in my view.

I still find it disappointing that  the LDM team won't follow Java's lead here and properly deprecate `_` as a valid identifier. Sure it'll break some folk's code, but a simple code fix that replaced them with `@_` would quickly patch fix such code until folk had time to properly fix those names.

**Comment by tpetrina** [comment]
@DavidArno Lots of stuff are weird in this world, like DateTime formatting in US and people still do it.
For example, I would prefer:

```csharp
public static bool operator ==(Unit, Unit) => true;
public static bool operator !=(Unit, Unit) => false;
```

**Comment by jnm2** [comment]
@tpetrina When would those operators ever be useful? They only come into play when comparing two instances which are statically known to be Unit, so the the code that would have used `==` or `!=` should just remove the check. (Assuming you define Unit as an empty struct so that it doesn't have equality operators at all.)

**Comment by tpetrina** [comment]
That is not my code, see second post in the thread.

**Comment by DavidArno** [comment]
@jnm2, [see my comment above](https://github.com/dotnet/csharplang/issues/111#issuecomment-281079222). It's genuine code. Of course, whether they are needed as you say, is a whole different question. With v8 and nullable ref types, `null` will be able to replace `unit` in so many cases that the type itself might become unnecessary.

**Comment by jnm2** [comment]
@tpetrina My bad! @DavidArno the fact that it's returning a constant means that the programmer should know that anyway; since operators are not virtual, they already know that both operands are Units. `IEquatable<Unit>` is still super important though.

**Comment by DavidArno** [comment]
@OJacot-Descombes, the language team are really reluctant to even make it a warning. Good practice says "treat warnings as errors". So when they introduce a new warning, that becomes a breaking change for some. The team *really* hate breaking changes. They set the justification bar very high for letting such changes through.

I suspect at some stage they may accrue enough evidence that `_` as a discard is causing sufficient confusion to justify it (as they did with changing closure behaviour with foreach loops) , but we aren't there yet.

**Comment by CyrusNajmabadi** [comment]
> (as they did with changing closure behaviour with foreach loops) , 

I think those are very different scenarios.  The loop case is one where the language was not just behaving in a counterintuitive fashion, it was *actively* leading people to write code that looked correct but could be very wrong depending on the circumstance.

The same doesn't really apply for ```_```.  It's normally a discard, except in the case where you don't want it to be a discard because you're using it as a variable.  This duality doesn't really cause a problem with code.  For exapmle, if the code treats it as a discard that doesn't really break your code (since you weren't using it as a variable).  And if you were using it as a variable, preserving htat logic doesn't break people who want it to be a discard, etc. etc.

'foreach' was a *very* special case because of how actively broken reasonably looking code was.  Discard-confusion (if that really even amounts to anything) is practically nothing like that.

**Comment by CyrusNajmabadi** [comment]
>I still find it disappointing that the LDM team won't follow Java's lead here and properly deprecate _ as a valid identifier. Sure it'll break some folk's code, 

You've now broken people's code.  For what benefit?  What is the actual end value that is sufficiently positive to have people go through the pain of the break?  Why not have both worlds?  Provide the benefit, and **not** break people?

**Comment by DavidArno** [comment]
> You've now broken people's code. For what benefit?

Clarity. And clarity wins every time in my book.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> You've now broken people's code. For what benefit? What is the actual end value that is sufficiently positive to have people go through the pain of the break? Why not have both worlds? Provide the benefit, and not break people?

To resolve ambiguity.  Even if it's uncommon for `_` to be a field somewhere it's perfectly legal and now the language has a built-in trap to accidentally overwriting said field when someone wants to opt-in to using discards, made worse by the fact that the compiler *must* favor the previous behavior in order to preserve compatibility.  The compiler may know for a fact that there is no `_` in scope, but the developer doesn't without going out of their way, and the tooling doesn't even help as Visual Studio displays both in black by default.

I'm sorry, Java's approach was significantly better here.  Deprecate the identifier over two versions and finally remove entirely.  That's so much better than overloading two diametrically opposed concepts into a single character and trying to force that into the language.

**Comment by CyrusNajmabadi** [comment]
Clearly that's not hte case for many customers.  Why should those views win out over their prefernce to not be broken unnecessarily?  Note: this is not hypothetical.  Breaks happen all the time, and we do hear about it *directly* from people that are not at all happy that their code which has been working fine is now considered broken.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> Clearly that's not hte case for many customers. Why should those views win out over their prefernce to not be broken unnecessarily?

Break the few to benefit the many.  Java's marketshare didn't take a dive when they did this.  C# would've been just fine.  The handful of projects using this would be updated and everyone would've moved on.

**Comment by jcouv** [comment]
Filed https://github.com/dotnet/roslyn/issues/26594 for an analyzer that warns if you use `_` as an identifier.

**Comment by CyrusNajmabadi** [comment]
>  Java's marketshare didn't take a dive when they did this.

There are far to many factors to be able to judge any sort of causal relationship here.  For all we know they might have grown more and this was a limiting factor.  

I'm not saying that this could not be done.  I'm saying that the benefits seem absolutely miniscule and will just end up being a pain for people.

I mean, it's not like we don't have ample historical precedent here.  We added 'var' and 'nameof' and didn't deprecate, and the world has gone on *just fine*, despite the fact that some people **do** use 'var' as an actual type name.  The need to deprecate has not actually arisen, and it seems to only be language enthusiasts who seem to actually have a problem here.  In practice the issues appear non-existent.  Whereas, when breaks happen, there is enormous pain felt and we get that feedback *straight on* from large sectoins of customers hit by it.

**Comment by CyrusNajmabadi** [comment]
> This is why I'm comparing it to Java, not Swift. 

TypeScript is another example.  I've broken on nearly every update of that language.  (Haven't tried 2.8 yet, but i expect the same).  IMO, that's more ok for that language given their target audience, and the acceptance of breaks/churns/new-shiny every month or so.  C# and .Net target very different communities, including large organizations that are slower to adopt and are not as willing to take this sort of pain upon adoption.

**Comment by HaloFour** [comment]
@jcouv 

Might I suggest also adding it to warning waves, if that's ever to happen?

@CyrusNajmabadi 

> I mean, it's not like we don't have ample historical precedent here.

Definitely, and don't get me wrong, I generally applaud the backcompat efforts by the team.  In the majority of the cases if the compiler confuses an identifier with a contextual keyword the result is broken compilation.  I'm sure that someone could come up with pathological examples where this isn't the case, but they're probably exceptionally rare.  But in this very specific case mixing the two results in accidentally overwriting a variable or field unintentionally.  This is why I lobbied against using `_` as a discard/wildcard to begin with.

**Comment by CyrusNajmabadi** [comment]
> or field unintentionally

I would definitely love a real world example of someone actually having a mutable field called '_', and then intending to have a discard, but having that discard write into the field...

--

It seems like that case would be *much* less common or likely than jsut hte case of people using _ today and having that code now break.

**Comment by CyrusNajmabadi** [comment]
> Might I suggest also adding it to warning waves, if that's ever to happen?

Note: i love warning waves.  But they don't mean that things can actually be removed.  It just means the compiler can add new warnings in a way that is felt to eb safe (precisely because of hte back compat concerns).  People can opt into new warning waves and get the new warnings/errors, but the language still needs to preserve all the old semantics.

This allows teh customer to be the one that opts into choosing the time when the value of the warning outweighs the drawbacks of any breaks.  I'm hugely supportive of this as it puts the customer in control, while not preventing them from upgrading, instead of us deciding for everyone.

**Comment by CyrusNajmabadi** [comment]
Note: i'm going to bow out here (with regard to back-compat).  I think this horse has been beaten to death.  I think the teams' position on back compat is well understood, and i hope we can improve things here with tooling/warning-waves.  Repeating myself on these points probably won't help anything.

**Comment by DavidArno** [comment]
@CyrusNajmabadi, @jnm2,
> I would definitely love a real world example of someone actually having a mutable field called '_', and then intending to have a discard, but having that discard write into the field...

Well, guess what? 😀 

[It so happens that I have a property, `__`](https://github.com/DavidArno/SuccincT/blob/master/src/SuccincT/PatternMatchers/Any.cs). Now the thing is, it started life as `_` as it's used in pattern matching in SuccincT to represent "anything" (as per F#'s use if `_` in patterns). Then the team made `_` a discard. Not wishing to cause grief to users of SuccincT, I renamed it `__`. 

Of course, if I'd not been hanging around here, knowing about discards before they were released, then I could have unleashed this on folk. I think the one saving grace would have been that its not mutable. But it's pretty damned close to what you asked for...

**Comment by CyrusNajmabadi** [comment]
> But it's pretty damned close to what you asked for...

And yet not what i asked for :D

**Comment by DavidArno** [comment]
@CyrusNajmabadi, you are a hard man to please! 😆

**Comment by CyrusNajmabadi** [comment]
That is the warning wave proposal @firelizzard18 .  Note we would not make someting invalid on an upgrade scenario.  You should be able to move to newer versions of the langauge without breaking your code.  Code warnings/breaks should only come through a separate opt-in mechanism (hence warnings waves).

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi, you are a hard man to please! 😆

I think it's very worthwhile when discussing these sorts of 'issues' to strongly understand if they reflect real problems the ecosystem is having, or if they're 'warts' that some people don't like and want to be excised in the name of 'purity'.

To make a breaking change it *must* be the case that this is a significant problem impacting a lot of people.  That was one the case for 'foreach' capture.  It kept biting people all over the place and it was one of those cases that people felt must assuredly be a bug.  We ourselves could not justify the actual behavior of the language (beyond: well... we shipped that way).  Indeed, it was unclear if there was an actual use case out there where the current bahavior was actually desirable (no one coudl find one).  These are the 'perfect storm' circumstances that justify breaking changes.

**Comment by alrz** [comment]
>The C# team is pretty obsessed with backward compatibility, which is usually a great thing,

At this point, I think if we remove all the code that is merely there for backward-compat, roslyn repo would be reduced to about half of its size.

**Comment by DavidArno** [comment]
@bkoelman,

Those "standards" lost me on the first rule, "*Use US English*". I'll use the language appropriate to the audience, thanks very much.

**Comment by jnm2** [comment]
@DavidArno Isn't the audience the entire world?

**Comment by dennisdoomen** [comment]
> Those "standards" lost me on the first rule, "Use US English". I'll use the language appropriate to the audience, thanks very much.

In this context, the _audience_ are the other developers.

**Comment by DavidArno** [comment]
@Jnm2, propriety code makes up a significant part of all code written. Having a standard that only applies to code with a public audience seems ... shortsighted.

**Comment by jnm2** [comment]
@DavidArno I wouldn't want to put an additional barrier in the way of open-sourcing part of a proprietary codebase.

**Comment by DavidArno** [comment]
@jnm2, 

So you feel folk should learn English and only use the US variation of that language in their code regardless of their first language, just in case they want to open source it is future? That seems more than a little arrogant.

**Comment by CyrusNajmabadi** [comment]
this conversation seems non-constructive.

**Comment by jnm2** [comment]
@DavidArno No, I don't think I feel that way. @CyrusNajmabadi Sorry, that was me.

**Comment by sirius0xff** [comment]
One possible solution for the problem of breaking change is use _ as discard parameter for lambda only if _ is/are not referenced.

Which would allow both:

`SomeEvent += (_,_,_,_)=>foo();`
and
`SomeList.Select(_=>_.Foo).MoreLinq....`

And this would result in an error:
`SomeEvent += (_,_,_,_)=>_.bar();`

==========================================================
If nested, take the closest one in scope:
```
SomeEvent += (_,_,_,_)=>SomeList.Select(_=>_.Foo).MoreLinq....
//same as:
SomeEvent += (_,_,_,_)=>SomeList.Select(__=>__.Foo).MoreLinq....
```
or simply forbid it:
`SomeEvent += (_,_,_,_)=>SomeList.Select(_=>_.Foo).MoreLinq.... //error`

==========================================================
It may cause confusion tough...

**Comment by gapspt** [comment]
If only a special character was chosen as the discard, other than a valid identifier...
Something such as ? or * would work just fine...

**Comment by gafter** [comment]
Since discards were added in C# 7.0 (a few language versions ago) we are unlikely to change the character we use for them.

**Comment by gapspt** [comment]
It was more me venting than anything else...
But it actually would be possible to allow for both _ and a new special character, where the new character would allow for more possibilities, but the _ would keep retro-compatibility.

**Comment by YairHalberstadt** [comment]
@matkoch
That kind of goes against the point of discards though

**Comment by matkoch** [comment]
@YairHalberstadt that's why I started my statement like I did ;)

**Comment by YairHalberstadt** [comment]
Fair enough. @davidarno would hate it though 😉.

I prefer not allowing this kind of hack, but TBH I don't really care that much.

**Comment by alrz** [comment]
> ```cs
> DoSomething(_ => _
>    .SomethingNested(_ => _
>        .AndAnotherNesting()));
> ```

I think what you want is "parameter shadowing in lambdas" which is already being considered.
```cs
DoSomething(s => s
    .SomethingNested(s => s // shadows the previous `s`
        .AndAnotherNesting()));
```
afaik, for now it will be implemented only for local functions as a part of "static local functions" (https://github.com/dotnet/csharplang/issues/1565).

**Comment by dadhi** [comment]
Anyone knows if exist a similar proposal for Linq expressions `from` bindings?

**Comment by jeroen-mostert** [comment]
I'd like this not just for its own sake, but also so the compiler could be meaningfully augmented with a new warning: "unused parameter to lambda". I had a nasty bug in my code of the form 

    this.Bar = collection.Sum(c => Bar);

where I _intended_ to write

    this.Bar = collection.Sum(c => c.Bar);

Obvious when you see it, mysterious when you don't!

Of course a warning for unused parameters could be added without parameter discards, but it would be obnoxious enough as to be useless. (Realistically, even with discards such a warning probably can't be added in the core compiler anyway, simply because all pre-discard code isn't going away any time soon, but at least things like custom code analyzers could make hay of it for new code bases.)

**Comment by jeroen-mostert** [comment]
I like this not just for itself but also so code analyzers can meaningfully start to issue "lambda parameter unused" warnings about code like

    this.Bar = collection.Sum(c => Bar);

where of course the intent was to write

    this.Bar = collection.Sum(c => c.Bar);

(And yes, I came here from fixing just such a bug.) CA1801 or vanilla Roslyn don't complain about this, and indeed such a warning would not be very practical today since it'd have to be suppressed very often. (I don't know if Resharper can warn for it, I don't have that. Having discards available makes the warning more viable regardless of the analyzer.)

**Comment by yaakov-h** [comment]
@jeroen-mostert A single-parameter lambda couldn't become a discard for backwards compatibility. This proposal only treats _multiple_ parameters named `_` as discards.

**Comment by jeroen-mostert** [comment]
@yaakov-h: Understood, but if `_` becomes idiomatic as a discard for _all_ lambda parameters, then `_` can be treated as a logical discard by analyzers even for single-parameter lambdas, leaving only the (unusual and exceptional) case of `_` actually being used in the lambda body. (Which, realistically, analyzers can then *also* warn about as something you probably want to write differently.)

**Comment by jcouv** [comment]
I renamed the issue as it is being broadened from lambda parameters to any/all parameters.

**Comment by jcouv** [comment]
From discussion today in LDM, we're going to keep the feature narrow for now (only support it for lambdas and anonymous methods as initially proposed).

**Comment by YairHalberstadt** [comment]
Can I ask why doing this for all methods is more complicated?

**Comment by jcouv** [comment]
@YairHalberstadt It's not a question of implementation cost, but trade-off between language complexity and benefit to users. Methods are part of an API (they can be invoked, whereas lambdas can only be converted to delegates, so the parameters names are never accessible to callers).
It is possible that discards would be generalized to methods later on, depending on feedback. More notes will be posted for today's meeting.

**Comment by atifaziz** [comment]
> lambdas can only be converted to delegates, so the parameters names are never accessible to callers

Aren't they available through reflection? I rely on that in a library.

**Comment by jnm2** [comment]
@jcouv Can local functions be included? They are often converted to delegates.

**Comment by DavidArno** [comment]
@jcouv,

You raise a good point re methods that I hadn't thought of. Consider the following code example,
```cs
public interface I 
{
    public void M(int a, int b);
}

public class C : I
{
    public void M(int _, int _) {}
}

public static class P
{
    private static void Main()
    {
        I i = new C();
        C c = new C();
        
        i.M(a:1, b:1);
        c.M(_:1, _:1);
    }
}
```
In `C`, I've used discards for `M` as I don't use the parameters. This makes sense to me: why name a parameter that isn't used?


But if I'm a developer working in a team that insists on named parameters (not sure if such a team exists, but thinking worst case here), then that line `c.M(_:1, _:1);` can't compile as I've trying to name those discards.

My motivation for wanting discards in methods (and operators, though I'm just lumping them in with methods here) is for situations such as when defining event handlers, operators and the like. In these situations, I have to specify the parameters to meet the contract. But if I then don't use those parameters, the compiler complains. So I want to use discards there to keep the compiler happy. 

But maybe this is the wrong solution to the problem. If a method must have a parameter to meet the contract, but that parameter isn't used, the compiler puts the developer in an impossible situation. So maybe the simpler solution is to just not warn on those unused parameters?

**Comment by Joe4evr** [comment]
@DavidArno I think in the case you present, that should likely be a compiler error since `C.M` is an *implicit* implementation of `I.M` and there need to be unambiguous identifiers for the first and second parameter because `M` is a public API regardless of whether or not `C` happens to be followed by `: I`.

*But*, I think explicit implementations maybe could allow discarded parameters since those methods are *only* callable through the interface signature. A potential (but perhaps unlikely) addition to that would be to also allow it in method overrides provided that the overriding class is less visible than the base class? Idk, it's obviously quite a complicated scenario.

**Comment by HaloFour** [comment]
@DavidArno 

> So maybe the simpler solution is to just not warn on those unused parameters?

I agree, the better approach would seem to have the analyzers not generate diagnostics for unused parameters that are required to meet the contract, such as for implicit implementation of an interface member or for overloaded operators, etc.

In general I'm a little squeamish about extending `_` to all parameters.  Lambdas and local functions are at least internal implementation details, but public methods are more likely to be walked via reflection and I have concerns over how such parameters would be interpreted by tooling.

**Comment by Logerfo** [comment]
Including local functions sounds perfect. Even private methods have their reflection concerns.

**Comment by louthy** [comment]
Will this also work for LINQ (as LINQ gets compiles down to lambdas)?  For those of us that use LINQ for [more than just querying an enumerating](https://github.com/louthy/language-ext) this would be an absolute godsend. 

It would reduce the calls for an extension to the LINQ grammar.  For example, I'm just working on a new IO monad:
```c#
static SIO<Runtime, ConsoleKeyInfo> ShowError(Error err) =>
    from _1 in setColor(Red)
    from _2 in writeLine(err.Message)
    from _3 in writeLine()
    from _4 in writeLine(StackTrace(err))
    from _5 in setColor(White)
    from k in readKey
    select k;
```
I would prefer to do this:
```c#
static SIO<Runtime, ConsoleKeyInfo> ShowError(Error err) =>
    do setColor(Red)
    do writeLine(err.Message)
    do writeLine()
    do writeLine(StackTrace(err))
    do setColor(White)
    do readKey
    select k;
```
But would be fine with just this (as a stopgap until the day that LINQ gets some much needed love):
```c#
static SIO<Runtime, ConsoleKeyInfo> ShowError(Error err) =>
    from _ in setColor(Red)
    from _ in writeLine(err.Message)
    from _ in writeLine()
    from _ in writeLine(StackTrace(err))
    from _ in setColor(White)
    from k in readKey
    select k;
```

**Comment by ChaseFlorell** [comment]
I asked this question on Twitter, but figured I should bring it over here
https://twitter.com/ChaseFlorell/status/1286159277451653121

Question, for those insane folks who use `_ => {}` and expect a value from it, is this a breaking change?

Is there a different character you can use that would have otherwise been invalid pre C# 9? `. => {}` perhaps?

**Comment by jnm2** [comment]
@ChaseFlorell No, it is not a breaking change. This feature only kicks in when at least two parameters are discarded.

**Comment by CyrusNajmabadi** [comment]
@jcouv do we have an existing issue covering multiple discards for methods, or multiple discards for locals?

**Comment by jcouv** [comment]
@CyrusNajmabadi Tracked by https://github.com/dotnet/csharplang/issues/2180

**Comment by CyrusNajmabadi** [comment]
Thanks!
